lib/pengine/pe_digest.c

```diff
@@ -242,27 +242,28 @@ calculate_secure_digest(op_digest_cache_t *data, pe_resource_t *rsc,
         pcmk__xe_remove_matching_attrs(data->params_secure, attr_in_string,
                                        (void *) secure_list);
     }
-    if (pcmk_is_set(pcmk_get_ra_caps(class),
-                    pcmk_ra_cap_fence_params)) {
+    if (compare_version(op_version, "3.16.0") < 0
+        && pcmk_is_set(pcmk_get_ra_caps(class),
+                       pcmk_ra_cap_fence_params)) {
         /* For stonith resources, Pacemaker adds special parameters,
-         * but these are not listed in fence agent meta-data, so the
-         * controller will not hash them. That means we have to filter
-         * them out before calculating our hash for comparison.
+         * but these are not listed in fence agent meta-data, so with older
+         * versions of DC, the controller will not hash them. That means we have
+         * to filter them out before calculating our hash for comparison.
          */
         pcmk__xe_remove_matching_attrs(data->params_secure, is_fence_param,
                                        NULL);
     }
     pcmk__filter_op_for_digest(data->params_secure);
 
     /* CRM_meta_timeout *should* be part of a digest for recurring operations.
-     * However, currently the controller does not add timeout to secure digests,
-     * because it only includes parameters declared by the resource agent.
+     * However, with older versions of DC, the controller does not add timeout
+     * to secure digests, because it only includes parameters declared by the
+     * resource agent.
      * Remove any timeout that made it this far, to match.
-     *
-     * @TODO Update the controller to add the timeout (which will require
-     * bumping the feature set and checking that here).
      */
-    xml_remove_prop(data->params_secure, CRM_META "_" XML_ATTR_TIMEOUT);
+    if (compare_version(op_version, "3.16.0") < 0) {
+        xml_remove_prop(data->params_secure, CRM_META "_" XML_ATTR_TIMEOUT);
+    }
```

Comment:
we should also remove "attr" for older DCs

Reply:
It sounds even hackier ... And we don't know whether any other resource agents might support "attr" parameter too. Probably for old DC,  rather than using params_all, we should just keep the old way.

Reply:
> Whoops I meant "addr", not "attr" ...

Ah, interesting enough, I meant "addr" too :-)

Reply:
I'm wrong about that, nearly all of the bundle-\* tests use `addr="#uname#"` including bundle-nested-colocation, which does:

    (pe__add_bundle_remote_name@bundle.c:992) 	trace: Setting address for bundle connection rabbitmq-bundle-0 to bundle host overcloud-rabbit-0
    (pe__add_bundle_remote_name@bundle.c:992) 	trace: Setting address for bundle connection rabbitmq-bundle-1 to bundle host overcloud-rabbit-1
    (pe__add_bundle_remote_name@bundle.c:992) 	trace: Setting address for bundle connection rabbitmq-bundle-2 to bundle host overcloud-rabbit-2

Not sure if that helps anything though

Reply:
I'm pretty sure they are. The history entries do have digests. The CIB XML equivalent is generated by the scheduler each run and unpacked like any other resource.

Reply:
I mean in the specific test bundle-nested-colocation.xml, they are not existing in the CIB status of yet, and pcmk__handle_rsc_config_changes() handles data_set->input.  But pcmk__inject_action_result() will first time calculate the digests.

Reply:
Makes sense.

I set up a test cluster with a bundle running on a remote node. I actually did run into problems, but so far they appear to be a scheduler bug unrelated to this PR. I want to investigate it a bit more to be sure before merging.

Reply:
My problem did turn out to be an unrelated scheduler bug. Tricky one, too. If I can fix it reasonably soon, I'll do further testing on this, otherwise I'll just merge this and try to test more before the next release.

Reply:
No hurry. Take your time.